File: December 2024 Leetcode Solution/December-12.java
class Solution {
    public long pickGifts(int[] gifts, int k) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> b - a);
        for (int gift : gifts) {
            maxHeap.offer(gift);
        }
        
        for (int i = 0; i < k; i++) {
            int top = maxHeap.poll();
            maxHeap.offer((int)Math.sqrt(top));
        }
        
        long total = 0;
        while (!maxHeap.isEmpty()) {
            total += maxHeap.poll();
        }
        return total;
    }
}

File: December 2024 Leetcode Solution/December-12.cpp
class Solution {
public:
    long long pickGifts(std::vector<int>& gifts, int k) {
        std::priority_queue<int> maxHeap(gifts.begin(), gifts.end());
        for (int i = 0; i < k; ++i) {
            int top = maxHeap.top();
            maxHeap.pop();
            maxHeap.push((int)sqrt(top));
        }
        
        long long total = 0;
        while (!maxHeap.empty()) {
            total += maxHeap.top();
            maxHeap.pop();
        }
        return total;
    }
};
